src/csharp/Microsoft.Spark/Sql/ForeachWriter.cs

```diff
@@ -0,0 +1,184 @@
+    /// guarantees. However, if the streaming query is being executed in the continuous
+    /// mode, then this guarantee does not hold and therefore should not be used for
+    /// deduplication.
+    /// </description>
+    /// </item>
+    /// </list>
+    /// </para>
+    /// </remarks>
+    public interface IForeachWriter
+    {
+        /// <summary>
+        /// Called when starting to process one partition of new data in the executor.
+        /// </summary>
+        /// <param name="partitionId">The partition id.</param>
+        /// <param name="epochId">A unique id for data deduplication.</param>
+        /// <returns>True if successful, false otherwise.</returns>
+        bool Open(long partitionId, long epochId);
+
+        /// <summary>
+        /// Called to process each <see cref="Row"/> in the executor side. This method
+        /// will be called only if <c>Open</c> returns <c>true</c>.
+        /// </summary>
+        /// <param name="row">The row to process.</param>
+        void Process(Row row);
+
+        /// <summary>
+        /// Called when stopping to process one partition of new data in the executor side. This is
+        /// guaranteed to be called either <see cref="Open(long, long)"/> returns <c>true</c> or
+        /// <c>false</c>. However, <see cref="Close(Exception)"/> won't be called in the following
+        /// cases:
+        /// <list type="bullet">
+        /// <item>
+        /// <description>
+        /// CLR/JVM crashes without throwing a <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// <item>
+        /// <description>
+        /// <see cref="Open(long, long)"/> throws an <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="errorOrNull">
+        /// The <see cref="Exception"/> thrown during processing or null if there was no error.
+        /// </param>
+        void Close(Exception errorOrNull);
+    }
+
+    /// <summary>
+    /// Wraps a <see cref="IForeachWriter"/> and calls the appropriate methods as decribed in
+    /// the lifecycle documentation for the interface.
+    /// </summary>
+    internal class ForeachWriterWrapper
+    {
+        private readonly IForeachWriter _foreachWriter;
+
+        internal ForeachWriterWrapper(IForeachWriter foreachWriter) =>
+            _foreachWriter = foreachWriter;
+
+        internal IEnumerable<object> Execute(int partitionId, IEnumerable<Row> rows)
+        {
+            if (!TaskContextHolder.Get().LocalProperties.TryGetValue(
+                "streaming.sql.batchId",
+                out string epochIdStr) || !long.TryParse(epochIdStr, out long epochId))
+            {
+                throw new Exception("Could not get batch id from TaskContext");
+            }
+
+            Exception error = null;
+            bool opened = _foreachWriter.Open(partitionId, epochId);
```

Comment:
you should be getting a new (copied) instance of _foreachWriter right?

Reply:
A `ForeachWriterWrapper` object will be deserialized for every task.  Since the state of the `_foreachWriter` will be set during the initial `serialization`, then each task will have its own copy of `ForeachWriterWrapper` and by extension its own copy of `_foreachWriter`.